File: Gold이상문제_정리/도로포장_1162.java
package Gold이상문제_정리;

import java.util.*;
import java.io.*;

public class 도로포장_1162 {
	static int N, M, K;
	static boolean[][] visited;
	static ArrayList<City>[] adj;
	static long[][] dp; // N개의 노드들에 대해서 K번 포장하는 경우를 위한 dp배열
	static PriorityQueue<City> pq = new PriorityQueue<>(Comparator.comparingLong(s -> s.time));
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		K = Integer.parseInt(st.nextToken());

		visited = new boolean[N][K+1];
		dp = new long[N][K+1];
		for (int i = 0; i < N; i++)
			Arrays.fill(dp[i], Long.MAX_VALUE);
		adj = new ArrayList[N];
		for (int i = 0; i < N; i++)
			adj[i] = new ArrayList<>();

		for (int i = 0; i < M; i++)
		{
			st = new StringTokenizer(br.readLine());
			int s = Integer.parseInt(st.nextToken()) - 1;
			int e = Integer.parseInt(st.nextToken()) - 1;
			int t = Integer.parseInt(st.nextToken());
			adj[s].add(new City(e, t, 0));
			adj[e].add(new City(s, t, 0));
		}

		dp[0][0] = 0;
		pq.add(new City(0, 0, 0));
		solve();
		long ret = Long.MAX_VALUE;
		for (long time : dp[N-1])
			ret = Math.min(time, ret);
		System.out.println(ret);
	}

	private static void solve() {
		while (!pq.isEmpty())
		{
			City now = pq.poll();
			int wrapCnt = now.wrap;

			visited[now.end][wrapCnt] = true;

			if (dp[now.end][wrapCnt] < now.time) continue;

			dp[now.end][wrapCnt] = now.time; // 주어진 시간 비용
			for (City next : adj[now.end]) {
				// 하나의 다음 노드에 대해서 도로를 포장하는 경우 && 그냥 가는 경우 두개 다 계산 후 넣어준다

				if (!visited[next.end][wrapCnt]) {
					// 도로를 포장하는 경우
					if (wrapCnt < K && dp[now.end][wrapCnt] < dp[next.end][wrapCnt + 1]) {
						dp[next.end][wrapCnt + 1] = dp[now.end][wrapCnt];
						pq.add(new City(next.end, dp[next.end][wrapCnt + 1], wrapCnt + 1));
					}
					// 그냥 계산하는 경우
					if (dp[now.end][wrapCnt] + next.time < dp[next.end][wrapCnt]) {
						dp[next.end][wrapCnt] = dp[now.end][wrapCnt] + next.time;
						pq.add(new City(next.end, dp[next.end][wrapCnt], wrapCnt));
					}
				}
			}
		}
	}

	private static class City
	{
		int end;
		long time;
		int wrap;

		public City(int end, long time, int wrap) {
			this.end = end;
			this.time = time;
			this.wrap = wrap;
		}
	}
}

File: 다익스트라/도로포장_1162.kt
package 다익스트라

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import kotlin.math.min

/*
4 4 1
1 2 10
2 4 10
1 3 1
3 4 100
 */

private var N: Int = 0
private var M: Int = 0
private var K: Int = 0
private const val MAX = Long.MAX_VALUE
private lateinit var adj: Array<MutableList<Pair<Int, Int>>>
private lateinit var dp: Array<Array<Long>>
private lateinit var visited: Array<Array<Boolean>>

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val (n, m, k) = br.readLine().split(" ").map { it.toInt() }
    N = n
    M = m
    K = k

    adj = Array(N + 1) { mutableListOf() }
    dp = Array(N + 1) { Array(K + 1) { MAX } }
    visited = Array(N + 1) { Array(K + 1) { false } }

    repeat(M) {
        val (u, v, cost) = br.readLine().split(" ").map { it.toInt() }
        adj[u].add(v to cost)
        adj[v].add(u to cost)
    }

    dp[1][0] = 0
    dijkstra()
    var answer: Long = dp[N][0]
    for (i in 0..K) {
        answer = min(answer, dp[N][i])
    }

    print(answer)
}

private fun dijkstra() {
    val pq = PriorityQueue<Triple<Int, Long, Int>>(compareBy { it.second })
    pq.add(Triple(1, 0, 0))

    while (pq.isNotEmpty()) {

        val (cur, curCost, useBridge) = pq.poll()

        if (useBridge > K) continue

        if (dp[cur][useBridge] < curCost) continue

        visited[cur][useBridge] = true

        for ((next, nextCost) in adj[cur]) {
            if (!visited[next][useBridge]) {
                if (dp[cur][useBridge] + nextCost < dp[next][useBridge]) {
                    dp[next][useBridge] = dp[cur][useBridge] + nextCost
                    pq.add(Triple(next, dp[next][useBridge], useBridge))
                }
                if (useBridge < K && dp[cur][useBridge] < dp[next][useBridge + 1]) {
                    dp[next][useBridge + 1] = dp[cur][useBridge]
                    pq.add(Triple(next, dp[next][useBridge + 1], useBridge + 1))
                }
            }
        }
    }
}